scaleway/helpers_container.go

```diff
@@ -0,0 +1,233 @@
+	meta := m.(*Meta)
+	api := container.NewAPI(meta.scwClient)
+
+	region, err := extractRegion(d, meta)
+	if err != nil {
+		return nil, "", err
+	}
+	return api, region, nil
+}
+
+func newContainerHandler(ctx context.Context, d *schema.ResourceData, m interface{}) (*containerHandler, diag.Diagnostics) {
+	api, region, err := containerAPIWithRegion(d, m)
+	if err != nil {
+		return nil, diag.Errorf("unexpected container handler: %s", err)
+	}
+	return &containerHandler{ctx: ctx, api: api, region: region}, nil
+}
+
+// expandLbACL transforms a state acl to an api one.
+func expandMapStr(data interface{}) *map[string]string {
+	res := make(map[string]string)
+	raw := data.(map[string]interface{})
+	for k, v := range raw {
+		res[*expandStringPtr(k)] = *expandStringPtr(v)
+	}
+
+	return &res
+}
+
+func (ch *containerHandler) setCreateContainerRequest(d *schema.ResourceData) *container.CreateContainerRequest {
+	// required
+	name := d.Get("name")
+	namespaceID := d.Get("namespace_id")
+	privacyType := d.Get("privacy") // default unknown_privacy
+	protocol := d.Get("protocol")   // default unknown_protocol
+
+	req := &container.CreateContainerRequest{
+		Region:      ch.region,
+		NamespaceID: expandID(namespaceID),
+		Name:        *expandStringPtr(name),
+		Privacy:     container.ContainerPrivacy(privacyType.(string)),
+		Protocol:    container.ContainerProtocol(*expandStringPtr(protocol)),
+	}
+
+	// optional
+	if envVariablesRaw, ok := d.GetOk("environment_variables"); ok {
+		req.EnvironmentVariables = expandMapStr(envVariablesRaw)
+	}
+
+	if minScale, ok := d.GetOk("min_scale"); ok {
+		req.MinScale = scw.Uint32Ptr(uint32(minScale.(int)))
+	}
+
+	if maxScale, ok := d.GetOk("max_scale"); ok {
+		req.MaxScale = scw.Uint32Ptr(uint32(maxScale.(int)))
+	}
+
+	if memoryLimit, ok := d.GetOk("memory_limit"); ok {
+		req.MemoryLimit = scw.Uint32Ptr(uint32(memoryLimit.(int)))
+	}
+
+	if timeout, ok := d.GetOk("timeout"); ok {
+		req.Timeout = &scw.Duration{Seconds: timeout.(int64)}
+	}
+
+	if description, ok := d.GetOk("description"); ok {
+		req.Description = expandStringPtr(description)
+	}
+
+	if registryImage, ok := d.GetOk("registry_image"); ok {
+		req.RegistryImage = expandStringPtr(registryImage)
+	}
+
+	if maxConcurrency, ok := d.GetOk("max_concurrency"); ok {
+		req.MaxConcurrency = scw.Uint32Ptr(uint32(maxConcurrency.(int)))
+	}
+
+	if domainName, ok := d.GetOk("domain_name"); ok {
+		req.DomainName = expandStringPtr(domainName)
+	}
+
+	return req
+}
+func (ch *containerHandler) waitForContainerCreation(d *schema.ResourceData) (*container.Container, diag.Diagnostics) {
+	req := ch.setCreateContainerRequest(d)
+	res, err := ch.api.CreateContainer(req, scw.WithContext(ch.ctx))
+	if err != nil {
+		return nil, diag.Errorf("unexpected waiting container error: %s", err)
+	}
+
+	return ch.waitForContainer(res.ID)
+}
+
+func (ch *containerHandler) waitForContainer(containerID string) (*container.Container, diag.Diagnostics) {
```

Comment:
Avoid this kind of patterns that make it more difficult to read waiting operation.